[WIP] feature: Make filesystem cache table aware

RatulDawar opened the pull request:
Related Issue : https://github.com/trinodb/trino/issues/27276 


Earlier there was no provision for selective table caching, all the queried files from all the tables competed for LRU eviction, this PR introduces ability to cache specific tables by passing the table names in description. 
Setting this property for a particular table will led to Filesystem caching being enabled for that table itself, users can then check their S3 API logs to optimize their API call cost / performance as they will have business context around what tables contribute most to the API cost. 
Example - A `fact` table can be a highly queried fast changing table, while a `dim` table will be equally highly queried slowly changing table, so it will be more beneficial to cache the `dim` table, rather than the `fact` table. 

Introduces `fs.cache.include-tables` configuration property that accepts a list of table patterns to include in caching
   - Supports wildcards: `schema.table`, `schema.*`, or `*` to cache all tables
   - Default value is `['*']` (cache all tables) to maintain backward compatibility



Testing 
- [ ]  Tested locally only using minio. Will further test on production workload too.

RatulDawar committed:
Make filesystem cache table aware

RatulDawar committed:
Add dependency injection for FileSystemConfig in TableCachingPredicate

RatulDawar committed:
Add missing overrides

RotRotAl reviewed (changes requested):
Hi please fix your code (5 test shave failed), and submit signed cla as the bot commented.

RatulDawar committed:
Address review comments: Update ConfigDescription and javadoc wording

RatulDawar committed:
Update TestFileSystemConfig to include cacheIncludeTables property

findepi commented on lib/trino-filesystem-manager/src/main/java/io/trino/filesystem/manager/TableCachingPredicate.java:
```diff
@@ -0,0 +1,80 @@
+        return tables.stream()
+                .map(TableCachingPredicate::parseTableName)
+                .map(prefix -> (Predicate<SchemaTableName>) prefix::matches)
+                .reduce(Predicate::or)
```
Looks fancy, but it would be better to just collect parsed table names on a field
and implement predicate with explicit for loop over them.

also, if any entry is `*`, then other elements still should be parsed (validated), but don't need to be remembered

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergFileSystemFactory.java:
```diff
@@ -14,11 +14,25 @@
+     * @param cachingEnabled whether file system caching should be enabled
+     * @return a TrinoFileSystem instance
+     */
+    default TrinoFileSystem create(ConnectorSession session, Map<String, String> fileIoProperties, boolean cachingEnabled)
```
the problem is that `cachingEnabled = true` does not imply there will be any caching
it depends what the backing `TrinoFileSystemFactory` is. i.e. caching needs to be allowed on the FS layer as well

RatulDawar replied:
Seems fair, I think then it's make it virtually impossible to use apis with caching flags. 
Should we go guice dependencies ? That seems fine over handling this in cachekeyprovider.

findepi commented on lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystemFactory.java:
```diff
@@ -24,4 +24,14 @@ default TrinoFileSystem create(ConnectorSession session)
+    default TrinoFileSystem create(ConnectorIdentity identity, boolean cachingEnabled)
+    {
+        return create(identity);
+    }
```
if we want to go this direction with new methods in TFSF, please open a separate PR adding those methods.
We can add usage of this methods in FileMetastore (controlled by configuration). This will be a simpler prep change & will ensure we have agreement on the API.

RatulDawar replied:
@findepi I think these API's make this bypassing more robust both the Guice binding and api's with caching flag seems better approches than the current cache key provider with context internal file paths. 
I would prefer API changes given that it is more flexible for possible future requirements. Open to discussion on this... !

RatulDawar committed:
Initialize cacheIncludeTables to ['*'] and add validation for empty list

- Change default value from empty list to ImmutableList.of('*') to maintain backward compatibility
- Add @NotEmpty validation to reject empty list configuration
- Add test to verify empty list validation fails
- Add jakarta.validation-api and io.airlift:testing dependencies to pom.xml

RatulDawar committed:
Refactor TableCachingPredicate constructor to streamline predicate initialization and improve readability

RatulDawar commented on lib/trino-filesystem-manager/src/main/java/io/trino/filesystem/manager/TableCachingPredicate.java:
```diff
@@ -0,0 +1,80 @@
+    private final Predicate<SchemaTableName> predicate;
+
+    @Inject
+    public TableCachingPredicate(FileSystemConfig config)
```
Let me confirm if this is possible or not

RoeyoOgen commented:
@RatulDawar any update?

RatulDawar committed:
Add AssertJ dependency and implement TableCachingPredicate tests

- Added AssertJ as a test dependency in pom.xml for improved assertions.
- Created TestTableCachingPredicate to validate various matching scenarios for TableCachingPredicate.
- Introduced TestTableCachingIntegration to test caching behavior with schema and table patterns.

RatulDawar committed:
Enhance TrinoFileSystemFactory with caching control methods and add test for cache disabled scenario

- Added methods to TrinoFileSystemFactory for creating file systems with explicit caching control.
- Updated TestCacheFileSystemAccessOperations to include a test for scenarios where caching is disabled.
- Improved documentation for caching-related configurations in FileSystemConfig and TableCachingPredicate.
